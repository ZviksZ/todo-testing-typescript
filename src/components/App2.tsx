import * as React                                from 'react';
import Counter from "./Counter";

function App2() {
   return (
      <div className="App">
         <Counter />
      </div>
   );
}

export default App2;
