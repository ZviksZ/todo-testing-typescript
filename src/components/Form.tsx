import * as React   from "react";
import { useState } from "react";

export const Form: React.FC = () => {
  const [input, setInput] = useState("");

  const submit = e => {

  }

  return (
    <form onSubmit={submit}>
       <span data-testid="change-input-greeting">
        Welcome, {input === "" ? "Anonymous User" : input}!
      </span>
       <br />
      <input
        type="text"
        aria-label="user-name"
        placeholder="Your name"
        name="input"
        onChange={(e) => setInput(e.target.value)}
        value={input}
      />
    </form>
  );
};
