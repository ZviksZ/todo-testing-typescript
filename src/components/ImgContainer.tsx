import * as React from 'react';
import {useState} from "react";

interface Props {
   imgLink: string
   callback?: () => void
}

const mock = () => {}

export const ImgContainer: React.FC<Props> = ({imgLink, callback = mock}) => {
   const [show, setShow] = useState<Boolean>(false)

   const clickHandler = () => {
      setShow(prevState => !prevState)
      callback()
   }

   return (
      <div className="container">
         <img src={imgLink} alt={imgLink}/>
         {show &&  <img src={imgLink} alt={imgLink}/>}
         <button onClick={clickHandler}>Add img</button>
      </div>
   );
}

