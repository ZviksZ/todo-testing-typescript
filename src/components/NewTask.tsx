import * as React from 'react';

import {useContext} from "react";
import {ContextApp} from "./App";

import {ActionType, TaskName} from "../types/stateType";

const NewTask: React.FC = () => {
// получаем state и dispatch-метод
   const {state, changeState} = useContext(ContextApp);

// отправляем два действия редьюсеру todoReducer - добавление задачи и изменение инпута. После их успешной обработки переменная state обновится. Для уточнения интерфейса передаваемого события можно воспользоваться расширенными React-интерфейсами
   const addTask = (event: React.FormEvent<HTMLFormElement>, task: TaskName) => {
      event.preventDefault();
      changeState({type: ActionType.Add, payload: task})
      changeState({type: ActionType.Change, payload: ''})
   }

// аналогично - отправим изменение значения в инпуте
   const changeTask = (event: React.ChangeEvent<HTMLInputElement>) => {
      changeState({type: ActionType.Change, payload: event.target.value})
   }

   return (
      <>
         <form onSubmit={(event)=>addTask(event, state.newTask)}>
            <input type='text' onChange={(event)=>changeTask(event)} value={state.newTask}/>
            <button type="submit">Add a task</button>
         </form>
      </>
   )
};

export default NewTask;

