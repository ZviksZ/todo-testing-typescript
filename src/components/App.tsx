import * as React                                from 'react';
import {useReducer}                              from 'react';
import {Action, ActionType, ContextState, State} from "../types/stateType";
import NewTask                                   from "./NewTask";
import TasksList                                 from "./TasksList";
import {todoReducer}                             from "../reducers/todoReducer";

// Начальные значения стейта
export const initialState: State = {
   newTask: '',
   tasks: []
}

// <Partial> позволяет создать контекст без дефолтных значений
export const ContextApp = React.createContext<Partial<ContextState>>({});



const App: React.FC = () => {
// Используем созданный todoReducer, передав его аргументом в useReduser. Изначально в стейт попадёт initialState, и далее при диспатче (changeState) будет обновляться.
   const [state, changeState] = useReducer<React.Reducer<State, Action>>(todoReducer, initialState);

   const ContextState: ContextState = {
      state,
      changeState
   };

// Передаём в контекст результаты работы useReducer - стейт и метод его изменения
   return (
      <>
         <ContextApp.Provider value={ContextState}>
            <NewTask/>
            <TasksList/>
         </ContextApp.Provider>
      </>
   )
}

export default App;
