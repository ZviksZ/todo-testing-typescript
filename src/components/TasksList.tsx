import * as React from 'react';

import {ActionType, Task} from "../types/stateType";
import {useContext} from "react";
import {ContextApp} from "./App";

const TasksList: React.FC = () => {
// Получаем состояние и диспатч (названный changeState)
   const {state, changeState} = useContext(ContextApp);

   const removeTask = (taskForRemoving: Task) => {
      changeState({type: ActionType.Remove, payload: taskForRemoving})
   }
   const toggleReadiness = (taskForChange: Task) => {
      changeState({type: ActionType.Toggle, payload: taskForChange})
   }

   return (
      <>
         <ul>
            {state.tasks.map((task,i)=>(
               <li key={i} className={task.isDone ? 'ready' : null}>
                  <label>
                     <input type="checkbox" onChange={()=>toggleReadiness(task)} checked={task.isDone}/>
                  </label>
                  <div className="task-name">
                     {task.name}
                  </div>
                  <button className='remove-button' onClick={()=>removeTask(task)}>
                     X
                  </button>
               </li>
            ))}
         </ul>
      </>
   )
};

export default TasksList;
