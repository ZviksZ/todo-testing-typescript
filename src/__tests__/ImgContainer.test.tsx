import * as React                   from 'react';
import {shallow}                    from 'enzyme';
import {fireEvent, render, cleanup} from "@testing-library/react";
import {ImgContainer}               from "../components/ImgContainer";

let defImg = './defImg.jpg'

describe('<ImgContainer />', () => {
   afterEach(cleanup);

   it('hasn`t got changes', () => {

      const component = shallow(<ImgContainer imgLink={defImg} />);

      expect(component).toMatchSnapshot();
   });

   it('should be one img',  async () => {

      const component = shallow(<ImgContainer imgLink={defImg} />);


      const img = component.find("img");

      expect(img).toHaveLength(1);
   });

   it('should be two img',  async () => {
      const { container } = render(<ImgContainer imgLink={defImg} />);

      fireEvent.click(container.querySelector('button'))

      const img = container.querySelectorAll("img");

      expect(img).toHaveLength(2);
   });

   it('should btn be clicked',  async () => {
      const mockCallBack = jest.fn();
      const { container } = render(<ImgContainer imgLink={defImg} callback={mockCallBack}/>);


      expect(mockCallBack.mock.calls.length).toBe(0);

      fireEvent.click(container.querySelector('button'))

      expect(mockCallBack.mock.calls.length).toBe(1);
   });


})
