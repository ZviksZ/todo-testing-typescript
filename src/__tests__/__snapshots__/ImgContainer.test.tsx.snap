// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImgContainer /> hasn\`t got changes 1`] = `
<div
  className="container"
>
  <img
    alt="./defImg.jpg"
    src="./defImg.jpg"
  />
  <button
    onClick={[Function]}
  >
    Add img
  </button>
</div>
`;
